habitat-lab/habitat/tasks/rearrange/utils.py

```diff
@@ -785,3 +606,197 @@ def get_camera_lookat_relative_to_vertial_line(
+
+            if (
+                embodiement_heuristic_offsets is None
+                and agent_embodiement is None
+            ):
+                # No embodiement for collision detection, so return closest unocluded point
+                return batch_sample[0], facing_target_angle, True
+
+            # get orientation noise offset
+            orientation_noise_samples = []
+            if orientation_noise > 0 and max_orientation_samples > 0:
+                orientation_noise_samples = [
+                    np.random.normal(0.0, orientation_noise)
+                    for _ in range(max_orientation_samples)
+                ]
+            # last one is always no-noise to check forward-facing
+            orientation_noise_samples.append(0)
+
+            for orientation_noise_sample in orientation_noise_samples:
+                desired_angle = facing_target_angle + orientation_noise_sample
+                if embodiement_heuristic_offsets is not None:
+                    # local 2d point rotation
+                    rotation_2d = mn.Matrix3.rotation(-mn.Rad(desired_angle))
+                    transformed_offsets_2d = [
+                        rotation_2d.transform_vector(xz)
+                        for xz in embodiement_heuristic_offsets
+                    ]
+
+                    # translation to global 3D points at navmesh height
+                    offsets_3d = [
+                        np.array(
+                            [
+                                transformed_offset_2d[0],
+                                0,
+                                transformed_offset_2d[1],
+                            ]
+                        )
+                        + batch_sample[0]
+                        for transformed_offset_2d in transformed_offsets_2d
+                    ]
+
+                    if data_out is not None:
+                        data_out["offsets_3d"] = offsets_3d
+
+                    # check for offset navigability
+                    is_collision = False
+                    for offset_point in offsets_3d:
+                        if not (
+                            sim.pathfinder.is_navigable(offset_point)
+                            and (
+                                island_id == -1
+                                or sim.pathfinder.get_island(offset_point)
+                                == island_id
+                            )
+                        ):
+                            is_collision = True
+                            break
+
+                    # if this sample is invalid, try the next
+                    if is_collision:
+                        continue
+
+                if agent_embodiement is not None:
+                    # contact testing with collision shapes
+                    start_position = agent_embodiement.base_pos
+                    start_rotation = agent_embodiement.base_rot
+
+                    agent_embodiement.base_pos = batch_sample[0]
```

Comment:
There is a case that once ` agent_embodiement.base_pos` is assigned to `nan` point, we cannot reset the agent pose by giving a valid point. do we want to add a check for not assigning nan point?